Added EKS install docs and updated node version in docs build (#1042)
k8ssandra/k8ssandra-operator: docs/content/en/install/eks/_index.md

---
title: "Install K8ssandra on EKS"
linkTitle: "Amazon EKS"
weight: 2
description: >
  Complete **production** ready environment of K8ssandra on Amazon Elastic Kubernetes Service (EKS).
---

> Amazon [Elastic Kubernetes Service](https://aws.amazon.com/eks/features/) or "EKS" is a managed Kubernetes service that makes it easy for you to run Kubernetes on AWS and on-premises. EKS is certified Kubernetes conformant, so existing applications that run on upstream Kubernetes are compatible with EKS. AWS automatically manages the availability and scalability of the Kubernetes control plane nodes responsible scheduling containers, managing the availability of applications, storing cluster data, and other key tasks.

Reviewer: ```suggestion
Amazon [Elastic Kubernetes Service](https://aws.amazon.com/eks/features/) or "EKS" is a managed Kubernetes service that makes it easy for you to run Kubernetes on AWS and on-premises. EKS is certified Kubernetes conformant, meaning that existing applications running on upstream Kubernetes are compatible with EKS. AWS automatically manages the availability and scalability of the Kubernetes control plane nodes responsible for key tasks such as scheduling containers, managing the availability of applications, and storing cluster data.
```

{{% alert title="Tip" color="success" %}}

> Also available in followup topics are post-install steps and role-based considerations for [developers]({{< relref "/quickstarts/developer">}}) or [site reliability engineers]({{< relref "/quickstarts/site-reliability-engineer">}}) (SREs).

Reviewer: Do these post-install and consideration topics exist such that they could be linked to?

Reviewer: I would comment this out until you have substantiating doc to link to... otherwise it is just frustrating to the user.

{{% /alert %}}

## Requirements

> 1. Amazon EKS Cluster - Follow the [Amazon EKS Getting Started Guide](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html)
>    a. K8ssandra has been tested with `Managed nodes - Linux` instances. `Fargate - Linux` _may_ work, but has not been validated or tested. If you are interested in support for Fargate, please [open an issue](https://github.com/k8ssandra/k8ssandra-operator/issues/new?assignees=&labels=enhancement&projects=&template=feature_request.md&title=Support+for+Fargate+on+EKS).
>    b. IAM Role attached to EKS EC2 instances _must_ include permissions to Amazon S3 APIs if you are going to be leveraging backup and restore functionality. See this [IAM policy](https://github.com/k8ssandra/k8ssandra-terraform/blob/ca5b1038d8cee003c0732d5c93ae06592e0a8535/aws/scripts/policy_document.json#L159-L190) as an example. _Note your policy should limit requests to specific buckets and operations._
> 2. [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/) installed and configured to access the cluster
> 3. _Optional_ [Helm](https://helm.sh/docs/intro/install/) installed and configured to access the cluster
> 4. _Optional_ [Kustomize](https://kubectl.docs.kubernetes.io/installation/kustomize/) installed and configured to access the cluster

Reviewer: ```suggestion
1. Amazon EKS Cluster - Follow the [Amazon EKS Getting Started Guide](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-console.html).
   a. K8ssandra is tested with `Managed nodes - Linux` instances. `Fargate - Linux` _may_ work, but has not been validated or tested. If you are interested in support for Fargate, please [open an issue](https://github.com/k8ssandra/k8ssandra-operator/issues/new?assignees=&labels=enhancement&projects=&template=feature_request.md&title=Support+for+Fargate+on+EKS).
   b. IAM Role attached to EKS EC2 instances _must_ include permissions to Amazon S3 APIs if you are going to be leveraging backup and restore functionality. See this [IAM policy](https://github.com/k8ssandra/k8ssandra-terraform/blob/ca5b1038d8cee003c0732d5c93ae06592e0a8535/aws/scripts/policy_document.json#L159-L190) as an example. _Note your policy should limit requests to specific buckets and operations._
2. [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/) installed and configured to access the cluster.
3. _Optional_ [Helm](https://helm.sh/docs/intro/install/) installed and configured to access the cluster.
4. _Optional_ [Kustomize](https://kubectl.docs.kubernetes.io/installation/kustomize/) installed and configured to access the cluster.
```
If you are specifically talking about DataStax not validating or testing Fargate - Linux, you might want to say that.

Author: I'm alright with this wording regarding Fargate.

## Infrastructure Considerations

> Apache Cassandra can be deployed across a variety of [hardware profiles](https://cassandra.apache.org/doc/latest/cassandra/operating/hardware.html). Consider the hardware requirements of Apache Cassandra when defining EKS node groups. Determine if you are planning on running a single Cassandra instance per node or multiple instances per node. If you are planning on running multiple instances per node, you will need to ensure that the node has enough resources to support the number of instances you are planning on running. For example, if you are planning on running 3 Cassandra instances per node, you will need to ensure that the node has enough CPU, memory, and storage to support 3 Cassandra instances.

Reviewer: ```suggestion
Apache Cassandra can be deployed across a variety of [hardware profiles](https://cassandra.apache.org/doc/latest/cassandra/operating/hardware.html). Consider the hardware requirements of Apache Cassandra when defining EKS node groups. Determine if you are planning on running a single Cassandra instance per node or multiple instances per node. If the decision is to run multiple instances per node, ensure that the node has enough resources to support the number of instances that you plan to run. For example, to run three  Cassandra instances per node, the node must have enough CPU, memory, and storage to support three Cassandra instances.
```

## Dependencies

> 1. [Cert Manager](https://cert-manager.io/docs/installation/kubernetes/) - Cert Manager provides a common API for management of TLS certificates. It provides interfaces for communicating with a variety of Certificate Authority backends, including self-signed, Venafi, and Vault. K8ssandra Operator uses this API for certificates used by the various K8ssandra components.

Reviewer: ```suggestion
1. [Cert Manager](https://cert-manager.io/docs/installation/kubernetes/) - Cert Manager provides a common API for management of Transport Layer Security (TLS) certificates. Cert Manager's interfaces communicate with a variety of Certificate Authority backends, including self-signed, Venafi, and Vault. K8ssandra Operator uses this API for certificates used by the various K8ssandra components.
```

## Installation Type

> K8ssandra Operator may be deployed in one of two modes. `Control Plane` mode is the default method of installation. In this mode K8ssandra Operator watches for the creation and changes to `K8ssandraCluster` custom resources. When this mode is active it may create Cassandra resources within the local Kubernetes cluster **and / or** remote Kubernetes clusters (in the case of multi-region) deployments. When using K8ssandra Operator you should only have one instance running in `Control Plane` mode. Kubernetes clusters acting as remote regions for Cassandra deployments should be run in `Data Plane` mode. In this mode K8ssandra Operator will not directly reconcile `K8ssandraCluster` resources.

Reviewer: ```suggestion
K8ssandra Operator may be deployed in one of two modes. `Control Plane` mode is the default method of installation. A `Control Plane` instance of K8ssandra Operator watches for the creation and changes to `K8ssandraCluster` custom resources. When `Control Plane` is active Cassandra resources may be created within the local Kubernetes cluster **and / or** remote Kubernetes clusters (in the case of multi-region) deployments. When using K8ssandra Operator you should only have one instance running in `Control Plane` mode. Kubernetes clusters acting as remote regions for Cassandra deployments should be run in `Data Plane` mode. In `Data Plane` mode K8ssandra Operator does not directly reconcile `K8ssandraCluster` resources.
```
should -or must- have one instance of K8ssandra-operator in a Control-Plane node?

Author: Across all clusters where k8ssandra-operator is deployed one, and only one, must be configured in control plane mode.

How does this change your suggestion?

For example consider the following scenarios:

* Single region, shared-cluster - K8ssandra Operator is deployed in `Control Plane` mode on the cluster. It can manage the creation of `K8ssandraCluster` objects and reconcile them locally into Cassandra clusters and nodes.
* Single region, dedicated cluster roles - K8ssandra Operator is deployed in `Control Plane` mode on a small, dedicated, k8s cluster. A separate cluster running in `Data Plane` mode contains larger instances which will be used to run Cassandra. This creates a separation between roles for each cluster (management vs production).
* Multiple region, shared cluster - One region is configured with K8ssandra Operator running in `Control Plane` mode, this cluster will run both the management operators and Cassandra nodes. Additional regions are configured in `Data Plane` mode focusing solely on Cassandra nodes and services.

> * Multiple region, dedicated clusters - This model is similar to the last with the exception of a dedicated management cluster running in one of the regions. This follows the same pattern as the single region, dedicated cluster roles scenario.

Reviewer: ```suggestion
* Single-region, shared-cluster - K8ssandra Operator is deployed in `Control-Plane` mode on the cluster. It can manage the creation of `K8ssandraCluster` objects and reconcile them locally into Cassandra clusters and nodes.
* Single-region, dedicated cluster roles - K8ssandra Operator is deployed in `Control-Plane` mode on a small, dedicated, K8s cluster. A separate cluster running in `Data-Plane` mode contains larger instances in which to run Cassandra. This creates a separation between management vs. production roles for each cluster.
* Multiple-region, shared cluster - One region is configured with K8ssandra Operator running in `Control-Plane` mode. This cluster runs both the management operators and Cassandra nodes. Configure additional regions in `Data-Plane` mode to focus solely on Cassandra nodes and services.
* Multiple-region, dedicated clusters - This scenario is similar to a shared cluster with the exception of a dedicated management cluster running in one of the regions. This follows the same pattern as the single-region, dedicated cluster roles scenario.
"This follows the same pattern as the single-region, dedicated cluster roles scenario." this sentence seems superfluous and may be able to be deleted?

Reviewer: Line 41  "This follows the same pattern as the single-region, dedicated cluster roles scenario." this sentence seems superfluous and may be able to be deleted?

Author: > this sentence seems superfluous and may be able to be deleted?

That's probably fair, I was drawing a parallel, but it isn't really necessary.

Adding dashes between `Control` / `Data` and `Plane` differs from the introduction of the modes previously in the doc. We should be uniform in our presentation. Which one is appropriate so I don't make this mistake again 😁 ?

Reviewer: The dashes are uniformly used in the MC customer-facing doc and are used because they modify the noun 'plane'.

Author: Awesome, I wasn't indicating the suggestion was wrong. Rather that we must make sure it's applied uniformly across this page (and other pages) in the K8ssandra Operator docs. Would you like me to create a ticket indicating this is the standard to be used across our docs?

Reviewer: Sure! consistency would be good!

## Installation Method

K8ssandra Operator is available to install with either [Helm](https://helm.sh/) or [Kustomize](https://kustomize.io/). Both tools provide various pros and cons which are outside of the scope of this document. Follow the steps within the specific section that is applicable to your environment.

### Helm

> Helm is sometimes described as a package manager for Kubernetes. A specific installation of a Helm chart is called a release. There may be _multiple_ installations of a particular Helm chart within a Kubernetes cluster for the same package. This guide covers a single installation, but as a user you should be aware there may be multiple.

Reviewer: ```suggestion
Helm is sometimes described as a package manager for Kubernetes. A specific installation of a Helm chart is called a release. There may be _multiple_ installations of a particular Helm chart within a Kubernetes cluster for the same package. This guide covers a single installation.
```

#### Install Cert Manager
In addition to Helm, Cert Manager also provides installation support via `kubectl`. Install Cert Manager with:

```console
kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v1.9.1/cert-manager.yaml
```

#### Install K8ssandra Operator
Replace `X.X.X` with the appropriate [release](https://github.com/k8ssandra/k8ssandra-operator/releases) you want to install:

**Control Plane Mode:**

```bash
kustomize build "github.com/k8ssandra/k8ssandra-operator/config/deployments/control-plane?ref=vX.X.X" | kubectl apply --server-side -f -
```

**Data Plane Mode:**

```bash
kustomize build "github.com/k8ssandra/k8ssandra-operator/config/deployments/data-plane?ref=vX.X.X" | kubectl apply --server-side -f -
```

Verify that the following CRDs are installed:

* `certificaterequests.cert-manager.io`
* `certificates.cert-manager.io`
* `challenges.acme.cert-manager.io`
* `clientconfigs.config.k8ssandra.io`
* `clusterissuers.cert-manager.io`
* `issuers.cert-manager.io`
* `k8ssandraclusters.k8ssandra.io`
* `orders.acme.cert-manager.io`
* `reapers.reaper.k8ssandra.io`
* `replicatedsecrets.replication.k8ssandra.io`
* `stargates.stargate.k8ssandra.io`

> Check that there are two Deployments. The output should look similar to this:

Reviewer: ```suggestion
Check that there are two deployments. A sample result is:
```

```console
kubectl -n k8ssandra-operator get deployment
NAME                 READY   UP-TO-DATE   AVAILABLE   AGE
cass-operator-controller-manager   1/1     1            1           77s
k8ssandra-operator                 1/1     1            1           77s
```

Verify that the `K8SSANDRA_CONTROL_PLANE` environment variable is set to `true`:

```console
kubectl -n k8ssandra-operator get deployment k8ssandra-operator -o jsonpath='{.spec.template.spec.containers[0].env[?(@.name=="K8SSANDRA_CONTROL_PLANE")].value}'
```

## Next Steps

> With a complete installation you may now move forward with creating a `K8ssandraCluster` custom resource instance. See the [Quickstart Samples]({{< relref "quickstarts/samples" >}}) for a collection `K8ssandraCluster` resources that can be used as a starting point for your own deployment. Once a `K8ssandraCluster` is up and running consider following either the [developer]({{< relref "quickstarts/developer" >}}) or [site reliability engineer]({{< relref "quickstarts/site-reliability-engineer" >}}) quickstarts.

Reviewer: ```suggestion
Installation is complete.  You can proceed with creating a `K8ssandraCluster` custom resource instance. See the [Quickstart Samples]({{< relref "quickstarts/samples" >}}) for a collection of `K8ssandraCluster` resources to use as a starting point for your own deployment. Once a `K8ssandraCluster` is up and running consider following either the [developer]({{< relref "quickstarts/developer" >}}) or the [site reliability engineer]({{< relref "quickstarts/site-reliability-engineer" >}}) quickstarts.
```
